## Chapter 4 - Exercise 1: Thực hiện những yêu cầu liên quan đến series, và đối chiếu với kết quả cho trước

### Part 1: Thực hiện các phép toán trên series

In [1]:
import numpy as np
import pandas as pd

In [2]:
# Câu 1a: Cho arr_1 là mảng số nguyên chẵn [2, 4, 6, 8, 10], arr_2 là mảng số nguyên lẻ [1, 3, 5, 7, 11]
# Tạo biến kiểu Serries ser1 từ arr_1, ser2 từ arr_2
arr_1 = np.array([2, 4, 6, 8, 10])
arr_2 = np.array([1, 3, 5, 7, 11])


# In danh sách các phần tử của ser1 và ser2
ser1 = pd.Series(arr_1)
ser2 = pd.Series(arr_2)
print(ser1)
print(ser2)

0     2
1     4
2     6
3     8
4    10
dtype: int64
0     1
1     3
2     5
3     7
4    11
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     2
1     4
2     6
3     8
4    10
dtype: int32
0     1
1     3
2     5
3     7
4    11
dtype: int32
</pre>
  
</details>

In [3]:
# Câu 1b: Thực hiện phép toán và thể hiện kết quả của:  ser1 + ser2 
result = ser1 + ser2
result

0     3
1     7
2    11
3    15
4    21
dtype: int64

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     3
1     7
2    11
3    15
4    21
dtype: int32</pre>
  
</details>

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0    1
1    1
2    1
3    1
4   -1
dtype: int32</pre>
  
</details>

In [4]:
# Câu 1d: Thực hiện phép toán và thể hiện kết quả của:  ser1 * ser2
result = ser1 * ser2
result

0      2
1     12
2     30
3     56
4    110
dtype: int64

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0      2
1     12
2     30
3     56
4    110
dtype: int32</pre>
  
</details>

In [5]:
# Câu 1e: Thực hiện phép toán và thể hiện kết quả của:  ser1 / ser2
result = ser1 / ser2
result

0    2.000000
1    1.333333
2    1.200000
3    1.142857
4    0.909091
dtype: float64

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0    2.000000
1    1.333333
2    1.200000
3    1.142857
4    0.909091
dtype: float64</pre>
  
</details>

In [6]:
# Câu 2a: Kiểm tra xem các phần tử của ser1 có > các phần tử của ser2 không?
result = ser1 > ser2
result

0     True
1     True
2     True
3     True
4    False
dtype: bool

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     True
1     True
2     True
3     True
4    False
dtype: bool</pre>
  
</details>

In [7]:
# Câu 2b: Kiểm tra xem các phần tử của ser1 có < các phần tử của ser2 không?
result = ser1 < ser2
result

0    False
1    False
2    False
3    False
4     True
dtype: bool

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0    False
1    False
2    False
3    False
4     True
dtype: bool</pre>
  
</details>

In [8]:
# Câu 2c: Kiểm tra xem các phần tử của ser1 có = các phần tử của ser2 không?
result = ser1 == ser2
result

0    False
1    False
2    False
3    False
4    False
dtype: bool

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0    False
1    False
2    False
3    False
4    False
dtype: bool</pre>
  
</details>

In [9]:
# Câu 3a: Thêm 2 phần tử [6, 12] vào ser2
ser2 = ser2.append(pd.Series([6, 12]))
# In lại danh sách các phần tử của ser2
ser2

0     1
1     3
2     5
3     7
4    11
0     6
1    12
dtype: int64

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     1
1     3
2     5
3     7
4    11
0     6
1    12
dtype: int64</pre>
  
</details>

In [32]:
# Câu 3b: Tạo series ser3 chỉ chứa các phần tử có trong ser1 mà không có trong ser2. 
#         In danh sách các phần tử của ser3
only_in_ser1 = ser1.isin(ser2)
ser3 = ser1.get(only_in_ser1 == False, ser1)
ser3


0     2
1     4
3     8
4    10
dtype: int64

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     2
1     4
3     8
4    10
dtype: int32
</pre>
  
</details>

In [35]:
# Câu 3c: Tạo series ser4 chỉ chứa các phần tử có trong ser2 mà không có trong ser1. 
#         In danh sách các phần tử của ser4
only_in_ser2 = ser2.isin(ser1)
ser4 = ser2.get(only_in_ser2 == False, ser2)
ser4


0     1
1     3
2     5
3     7
4    11
1    12
dtype: int64

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     1
1     3
2     5
3     7
4    11
1    12
dtype: int64
</pre>
  
</details>

In [54]:
# Câu 4: Tạo series ser5 chứa các phần tử chỉ có trong ser1 và chỉ có trong ser2
#        In danh sách các phần tử của ser5
ser5 = ser1.append(ser2).unique()
ser5 = pd.Series(ser5)
ser5

0      2
1      4
2      6
3      8
4     10
5      1
6      3
7      5
8      7
9     11
10    12
dtype: int64

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     2
1     4
3     8
4    10
0     1
1     3
2     5
3     7
4    11
1    12
dtype: int64
</pre>
  
</details>

### Part 2: Truy xuất các phần tử, và thống kê thông tin trên series

In [63]:
# Câu 1a: Tạo series ser6 có 35 phần tử số nguyên ngẫu nhiên có giá trị trong khoảng từ 1 đến 9. 
np.random.seed(1)

# Cho biết kích thước (shape) của ser6
arr6 = np.random.randint(low = 1, high = 10, size = 35)
ser6 = pd.Series(arr6)
np.shape(ser6)

# Xem 5 dòng dữ liệu đầu tiên (head) và 5 dòng dữ liệu cuối cùng (tail) có trong ser6
print(ser6.head(5))
print(ser6.tail(5))

0    6
1    9
2    6
3    1
4    1
dtype: int64
30    8
31    4
32    7
33    6
34    2
dtype: int64


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>(35,)
0    6
1    9
2    6
3    1
4    1
dtype: int32
30    8
31    4
32    7
33    6
34    2
dtype: int32
</pre>
  
</details>

In [64]:
# Câu 1b: In danh sách các phần tử của ser6 theo dạng array
ser6.array

<PandasArray>
[6, 9, 6, 1, 1, 2, 8, 7, 3, 5, 6, 3, 5, 3, 5, 8, 8, 2, 8, 1, 7, 8, 7, 2, 1, 2,
 9, 9, 4, 9, 8, 4, 7, 6, 2]
Length: 35, dtype: int64

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>[6 9 6 1 1 2 8 7 3 5 6 3 5 3 5 8 8 2 8 1 7 8 7 2 1 2 9 9 4 9 8 4 7 6 2]
</pre>
  
</details>

In [65]:
# Câu 1c: Cho biết thông tin thống kê chung (describe()) của ser6
ser6.describe()

count    35.000000
mean      5.200000
std       2.763204
min       1.000000
25%       2.500000
50%       6.000000
75%       8.000000
max       9.000000
dtype: float64

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>count    35.000000
mean      5.200000
std       2.763204
min       1.000000
25%       2.500000
50%       6.000000
75%       8.000000
max       9.000000
dtype: float64</pre>
  
</details>

In [67]:
# Câu 1d: Cho biết tổng của các phần tử có trong ser6
ser6.sum()

182

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>182</pre>
  
</details>

In [80]:
# Câu 1e: Cho biết phần tử có tần suất xuất hiện nhiều nhất trong ser6
value_counts = ser6.value_counts()
value_counts

8    6
2    5
9    4
7    4
6    4
1    4
5    3
3    3
4    2
dtype: int64

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0    8
dtype: int32</pre>
  
</details>

In [100]:
# Câu 2:  Liệt kê các dòng trong ser6 mà giá trị chia hết cho 2 và cho 3
result = ser6.where((ser6 % 2 == 0) & (ser6 % 3 == 0)).dropna()
result

0     6.0
2     6.0
10    6.0
33    6.0
dtype: float64

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     6
2     6
10    6
33    6
dtype: int32</pre>
  
</details>

In [19]:
# Câu 3: In các phần tử ở vị trí 0, 5, 10, 15 trong ser6


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0     6
5     2
10    6
15    8
dtype: int32</pre>
  
</details>

In [20]:
# Câu 4: In ra các giá trị unique (array) trong ser6


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>array([6, 9, 1, 2, 8, 7, 3, 5, 4], dtype=int64)</pre>
  
</details>

In [21]:
# Câu 5: Tạo series ser7 với mỗi phần tử có giá trị = lập phương của phần tử trong ser6. 


# Xem 5 dòng dữ liệu đầu tiên (head) của ser7


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0    216.0
1    729.0
2    216.0
3      1.0
4      1.0
dtype: float64
</pre>
  
</details>

### Part 3: Tạo series từ list, chuỗi và biểu thức điều kiện 

In [22]:
# Câu 1: Cho list sau:
lst = ["abc", "defg", "htlmj", "dfg", "ljsac"]

In [23]:
# Câu 1a: Tạo series ser_chuoi từ lst 


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0      abc
1     defg
2    htlmj
3      dfg
4    ljsac
dtype: object
</pre>
  
</details>

In [24]:
# Câu 1b: Tạo series ser_dodai với mỗi phần tử có giá trị là chiều dài của mỗi phần tử trong ser_chuoi


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>0    3
1    4
2    5
3    3
4    5
dtype: int64</pre>
  
</details>

In [25]:
# Câu 2: Cho ser = pd.Series(np.array([1, 2, 4, 5, 8, 7, 6, 9])). 
# Sử dụng biểu thức điều kiện thích hợp để in ra các dòng trong ser có giá trị là số nguyên tố



<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>1    2
3    5
5    7
dtype: int32
</pre>
  
</details>

In [26]:
# Câu 3: Cho mẫu email như sau:
pattern ='[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,4}'

# Tạo một series ser_ch, với mỗi phần tử trong ser_ch là một chuỗi
# Gợi ý: 'reading newspaper from tuoitre.vn', 'tubirona@gmail.com', 'nguyen.nn@yahoo.com', 'tran_2014@hotmail.com.vn'


# In ra những dòng trong ser_ch thỏa điều kiện chuỗi là email


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>1          tubirona@gmail.com
2         nguyen.nn@yahoo.com
3    tran_2014@hotmail.com.vn
dtype: object</pre>
  
</details>

In [27]:
# Câu 4: Cho series:
ser_names = pd.Series(['Manufacturer', 'Model', 'CarType', 'Min_Price', 'Price', 'Max_Price',
       'MPG_city', 'MPG_highway', 'AirBags', 'DriveTrain', 'Cylinders',
       'EngineSize', 'Horsepower', 'RPM', 'Rev_per_mile', 'Man_trans_avail',
       'Fuel_tank_capacity', 'Passengers', 'Length', 'Wheelbase', 'Width',
       'Turn_circle', 'Rear_seat_room', 'Luggage_room', 'Weight', 'Origin',
       'Make'])
# Sử dụng biểu thức điều kiện thích hợp để in ra các dòng của ser_names thỏa điều kiện trong chuỗi có chữ 'Price'



<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>3    Min_Price
4        Price
5    Max_Price
dtype: object
</pre>
  
</details>